Services/DeviceModels.cs

```diff
@@ -81,19 +98,70 @@ public DeviceModel Get(string id)
             throw new ResourceNotFoundException();
         }
 
-        private List<string> GetDeviceModelFiles()
+        /// <summary>
+        /// Create a device model.
+        /// </summary>
+        public async Task<DeviceModel> InsertAsync(DeviceModel deviceModel)
+        {
+            if (this.CheckDeviceModelExistence(deviceModel.Id))
+            {
+                throw new ConflictingResourceException("Device model with id '" + deviceModel.Id + "' already exists!");
+            }
+
+            try
+            {
+                var result = await this.customDeviceModels.InsertAsync(deviceModel);
+                deviceModel.ETag = result.ETag;
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Unable to create a new device model", () => new { Exception = e });
+            }
+
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create or replace a device model.
+        /// </summary>
+        public async Task<DeviceModel> UpsertAsync(DeviceModel deviceModel)
         {
-            if (this.deviceModelFiles != null) return this.deviceModelFiles;
+            if (this.CheckDeviceModelExistence(deviceModel.Id))
+            {
+                this.log.Info("Unable to create a new device model; device model already exists.", () => new { deviceModel });
+                throw new ConflictingResourceException("Device model with id '" + deviceModel.Id + "' already exists!");
+            }
 
-            this.log.Debug("Device models folder", () => new { this.config.DeviceModelsFolder });
+            try
+            {
+                var result = await this.customDeviceModels.UpsertAsync(deviceModel);
+                deviceModel.ETag = result.ETag;
+            }
+            catch (Exception exception)
+            {
+                this.log.Error("Unable to update device model ", () => new { exception });
```

Comment:
same here, if UpsertAsync fails, is it ok to return a model at line 145? shouldn't some exception be handled in the caller?

Reply:
Done.